File: veriblock-core/src/main/java/org/veriblock/sdk/models/VeriBlockBlock.java
// VeriBlock Blockchain Project
// Copyright 2017-2018 VeriBlock, Inc
// Copyright 2018-2019 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.

package org.veriblock.sdk.models;

import org.veriblock.core.crypto.Sha256Hash;
import org.veriblock.core.utilities.Preconditions;
import org.veriblock.sdk.services.SerializeDeserializeService;

import java.util.Arrays;

public class VeriBlockBlock{
    private byte[] raw;
    private final int height;
    private final short version;
    private final VBlakeHash previousBlock;
    private final VBlakeHash previousKeystone;
    private final VBlakeHash secondPreviousKeystone;
    private final Sha256Hash merkleRoot;
    private final int timestamp;
    private final int difficulty;
    private final int nonce;
    
    private VBlakeHash hash;

    public int getHeight() {
        return height;
    }

    public short getVersion() {
        return version;
    }

    public VBlakeHash getPreviousBlock() {
        return previousBlock;
    }

    public VBlakeHash getPreviousKeystone() {
        return previousKeystone;
    }

    public VBlakeHash getSecondPreviousKeystone() {
        return secondPreviousKeystone;
    }

    public Sha256Hash getMerkleRoot() {
        return merkleRoot;
    }

    public int getTimestamp() {
        return timestamp;
    }

    public int getDifficulty() {
        return difficulty;
    }

    public int getNonce() {
        return nonce;
    }

    public byte[] getRaw() {
        return raw;
    }

    public VBlakeHash getHash() {
        return this.hash;
    }

    public VeriBlockBlock(int height,
                          short version,
                          VBlakeHash previousBlock,
                          VBlakeHash previousKeystone,
                          VBlakeHash secondPreviousKeystone,
                          Sha256Hash merkleRoot,
                          int timestamp,
                          int difficulty,
                          int nonce) {
        Preconditions.argument(previousBlock != null,
            "Previous block cannot be null");
        Preconditions.argument(previousBlock.length >= VBlakeHash.PREVIOUS_BLOCK_LENGTH,
            () -> "Invalid previous block: " + previousBlock.toString());
        Preconditions.argument(previousKeystone != null,
            "Previous keystone cannot be null");
        Preconditions.argument(previousKeystone.length >= VBlakeHash.PREVIOUS_KEYSTONE_LENGTH,
            () -> "Invalid previous keystone: " + previousKeystone.toString());
        Preconditions.argument(secondPreviousKeystone != null,
            "Second previous keystone cannot be null");
        Preconditions.argument(secondPreviousKeystone.length >= VBlakeHash.PREVIOUS_KEYSTONE_LENGTH,
            () -> "Invalid second previous keystone: " + secondPreviousKeystone.toString());
        Preconditions.argument(merkleRoot != null,
            "Merkle root cannot be null");
        Preconditions.argument(merkleRoot.length >= Sha256Hash.VERIBLOCK_MERKLE_ROOT_LENGTH,
            () -> "Invalid merkle root: " + merkleRoot.toString());

        this.height = height;
        this.version = version;
        this.previousBlock = previousBlock.trimToPreviousBlockSize();
        this.previousKeystone = previousKeystone.trimToPreviousKeystoneSize();
        this.secondPreviousKeystone = secondPreviousKeystone.trimToPreviousKeystoneSize();
        this.merkleRoot = merkleRoot.trim(Sha256Hash.VERIBLOCK_MERKLE_ROOT_LENGTH);
        this.timestamp = timestamp;
        this.difficulty = difficulty;
        this.nonce = nonce;

        this.raw = SerializeDeserializeService.serializeHeaders(this);
        this.hash = VBlakeHash.hash(this.raw);
    }

    public int getRoundIndex() {
        return getHeight() % Constants.KEYSTONE_INTERVAL;
    }

    public boolean isKeystone() {
        return getHeight() % Constants.KEYSTONE_INTERVAL == 0;
    }

    public VBlakeHash getEffectivePreviousKeystone() {
        if (getHeight() % Constants.KEYSTONE_INTERVAL == 1) return getPreviousBlock();

        return getPreviousKeystone();
    }

    @Override
    public boolean equals(Object o) {
        return this == o || o != null && getClass() == o.getClass() && Arrays.equals(SerializeDeserializeService.serialize(this), SerializeDeserializeService
            .serialize((VeriBlockBlock) o));

    }

}

File: veriblock-core/src/main/java/org/veriblock/sdk/models/VeriBlockBlock.kt
// VeriBlock Blockchain Project
// Copyright 2017-2018 VeriBlock, Inc
// Copyright 2018-2021 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.
package org.veriblock.sdk.models

import org.veriblock.core.bitcoinj.BitcoinUtilities
import org.veriblock.core.crypto.AnyVbkHash
import org.veriblock.core.crypto.PreviousBlockVbkHash
import org.veriblock.core.crypto.PreviousKeystoneVbkHash
import org.veriblock.core.crypto.VbkHash
import org.veriblock.core.crypto.MerkleRoot
import org.veriblock.core.crypto.TruncatedMerkleRoot
import org.veriblock.core.crypto.asVbkHash
import org.veriblock.core.utilities.BlockUtility
import org.veriblock.sdk.services.SerializeDeserializeService
import java.math.BigInteger
import java.util.Arrays

open class VeriBlockBlock(
    val height: Int,
    val version: Short,
    val previousBlock: PreviousBlockVbkHash,
    val previousKeystone: PreviousKeystoneVbkHash,
    val secondPreviousKeystone: PreviousKeystoneVbkHash,
    val merkleRoot: TruncatedMerkleRoot,
    val timestamp: Int,
    val difficulty: Int,
    var nonce: Long,
    private val precomputedHash: VbkHash? = null
) {
    val raw: ByteArray
        get() = SerializeDeserializeService.serializeHeaders(this)

    val hash: VbkHash by lazy {
        precomputedHash ?: BlockUtility.hashBlock(raw).asVbkHash()
    }

    fun getRoundIndex(): Int =
        height % Constants.KEYSTONE_INTERVAL

    fun isKeystone(): Boolean =
        height % Constants.KEYSTONE_INTERVAL == 0

    fun getEffectivePreviousKeystone(): AnyVbkHash =
        if (height % Constants.KEYSTONE_INTERVAL == 1) previousBlock else previousKeystone

    fun getDecodedDifficulty(): BigInteger = BitcoinUtilities.decodeCompactBits(this.difficulty.toLong())

    override fun equals(other: Any?): Boolean {
        return this === other || other != null && javaClass == other.javaClass && Arrays.equals(
            SerializeDeserializeService.serialize(this), SerializeDeserializeService
            .serialize(other as VeriBlockBlock)
        )
    }

    override fun toString(): String {
        return "$hash @ $height"
    }

    override fun hashCode(): Int {
        var result = precomputedHash?.hashCode() ?: 0
        result = 31 * result + height
        result = 31 * result + version
        result = 31 * result + previousBlock.hashCode()
        result = 31 * result + previousKeystone.hashCode()
        result = 31 * result + secondPreviousKeystone.hashCode()
        result = 31 * result + merkleRoot.hashCode()
        result = 31 * result + timestamp
        result = 31 * result + difficulty
        result = 31 * result + nonce.hashCode()
        return result
    }
}
